clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp

```diff
@@ -53,6 +57,140 @@ void TypeConversionsCheck::check(const MatchFinder::MatchResult &Result) {
+    } else {
+      handleImplicitCast(Result.Context, sourceType, destType,
+                         ImplicitCastDecl->getBeginLoc());
+    }
+  }
+
+  const auto *StaticCastDecl =
+      Result.Nodes.getNodeAs<CXXStaticCastExpr>(StaticCastId);
+  if (StaticCastDecl != nullptr) {
+    const QualType destType = StaticCastDecl->getType();
+    const QualType sourceType = StaticCastDecl->getSubExpr()->getType();
+    handleStaticCast(Result.Context, sourceType, destType,
+                     StaticCastDecl->getOperatorLoc());
+  }
+}
+
+bool TypeConversionsCheck::isSafeIntegralCast(clang::ASTContext *context,
+                                              QualType sourceType,
+                                              QualType destType) const {
+  if (sourceType->isIntegerType() && destType->isIntegerType()) {
+    // Check that the size of the destination integer is larger than source
+
+    const unsigned sourceBits = sourceType->isBitIntType()
+                                    ? context->getIntWidth(sourceType)
+                                    : context->getTypeSize(sourceType);
+    const unsigned destBits = destType->isBitIntType()
+                                  ? context->getIntWidth(destType)
+                                  : context->getTypeSize(destType);
+
+    // unsigned to signed conversion is fine if number of bits is strictly
+    // larger than the source, otherwise it may not fit
+    if ((destType->isSignedIntegerType() &&
+         sourceType->isUnsignedIntegerType()) &&
+        (destBits > sourceBits)) {
+      return true;
+    }
+
+    // If they are the same sign, they need to have at least the same number of
+    // bits
+    if ((destType->isSignedIntegerType() ==
+         sourceType->isSignedIntegerType()) &&
+        (destBits >= sourceBits)) {
+      // Explicit promotion of integer types is fine
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void TypeConversionsCheck::handleStaticCast(clang::ASTContext *context,
+                                            QualType sourceType,
+                                            QualType destType,
+                                            SourceLocation location) {
+  if (destType == sourceType) {
+    // If source and destination are the same type we can ignore the cast
+    return;
+  }
+
+  if (destType->isVoidType()) {
+    // Casting to void is always allowed
+    return;
+  }
+
+  if (isSafeIntegralCast(context, sourceType, destType)) {
+    // Integral promotion is allowed
+    return;
+  }
+
+  if (sourceType->isRealFloatingType() && destType->isIntegerType()) {
+    diag(location, "Floating point to integral conversion must use 'floor<T>', "
+                   "'ceil<T>' or 'round<T>'");
+    diag(location, "Conversion from %1 to %0", DiagnosticIDs::Note)
+        << destType << sourceType;
+    return;
+  }
+
+  if (sourceType->isIntegerType() && destType->isRealFloatingType()) {
+    // Allowed integral to floating point type
```

Comment:
although in principle there could be data loss here

Reply:
That is a fair  point indeed, but we should be able to make this conversion explicitly. Or would you prefer using a explicit function for these conversions (similar to `clipToFit<T>`)? I also understand that these methods would internally just use static cast and have a exception for this rule.

Reply:
static_cast is fine, but it means the "no data loss" check is not going to be fully automated and still needs to be checked for during review. Right?

Reply:
I would think so, yes. Or we can remove this and force the user to explain themselves by adding an explicit exception?

Reply:
Either works for me